ttnn/cpp/ttnn/operations/data_movement/repeat/repeat.cpp

```diff
@@ -36,11 +47,26 @@ ttnn::Tensor RepeatOperation::invoke(
                         (input_tensor.get_legacy_shape()[dim] * input_tensor.element_size()) % input_tensor.buffer()->alignment() == 0,
                         "Current repeat implementation requires aligned last dim when repeating on last dim");
                 }
-                output = operation::run_without_autoformat(RepeatDeviceOperation{dim, repeat_dims[dim], memory_config}, {output}).at(0);
+                auto outputs = operation::run_without_autoformat(RepeatDeviceOperation{dim, repeat_dims[dim], memory_config}, {output});
+                TT_FATAL(outputs.size() == 1, "ttnn.repeat: expected 1 output tensor from run_without_autoformat, but got {}", outputs.size());
+                output = outputs[0];
             }
             return {output};
-        }, {input_tensor}, output_tensors);
-    return output_tensors.at(0);
+        }, {formatted_input_tensor.has_value() ? *formatted_input_tensor : input_tensor}, output_tensors);
+    TT_FATAL(output_tensors.size() == 1, "ttnn.repeat: expected 1 output tensor, but got {}", output_tensors.size());
+    if (formatted_input_tensor.has_value()) {
+        auto zero_indices = std::vector<uint32_t>(input_tensor.get_legacy_shape().rank(), 0);
+        std::vector<uint32_t> end_indices(repeat_dims.size());
+        for (uint32_t i = 0; i < repeat_dims.size(); ++i) {
+            end_indices[i] = repeat_dims[i] * input_tensor.get_legacy_shape().without_padding()[i];
+        }
+        auto step = std::vector<uint32_t>(input_tensor.get_legacy_shape().rank(), 1);
+        auto sliced_output = ttnn::slice(output_tensors[0], zero_indices, end_indices, step, input_tensor.memory_config(), std::nullopt);
+        auto padded_to_tile = sliced_output.cpu().pad_to_tile(0.0).to(input_tensor.device(), input_tensor.memory_config());
```

Comment:
It would be amazing if we didn't need slicing and CPU padding. Can we try to find a solution to avoid those?

Reply:
The slice is essentially required for now as the device operation doesn't seem to be aware of padding. I think you may be right that the untilize you pointed out above is unnecessary, but I'll have to test it. If I can drop the utilize, I think I can also remove the trip back to cpu to add back in the padding, which would be a huge win indeed 😃 

In general, the TMs team will certainly loop back to this down the line and produce a better fix, but this is what I'm able to do with available bandwidth for now (not making changes to kernels, etc.) The issue prompting this change is a P0 so the focus is on unblocking the models team as fast as possible.

Reply:
See my latest commit. The slice is required for now, but I managed to delay the conversion to row major to just before the slice; slice doesn't seem to handle tiled padding well yet either. Without this, I hit FATALs due to the slice result not neatly fitting into a tile volume. There's also a whole lot of preprocessing required after the slice to re-tilize since we don't have the luxury of host-side pad_to_tile. I think it would be best to relocate this pad-to-tile logic to inside of ttnn::tilize. Thoughts @ntarafdar?